src/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ProjectSystem/UpToDate/BuildUpToDateCheckTests.cs

```diff
@@ -0,0 +1,461 @@
+            };
+        }
+
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public async Task IsUpToDateCheckEnabledAsync_DelegatesToProjectSystemOptions(bool isEnabled)
+        {
+            _isFastUpToDateCheckEnabled = isEnabled;
+
+            Assert.Equal(
+                isEnabled,
+                await _buildUpToDateCheck.IsUpToDateCheckEnabledAsync(CancellationToken.None));
+        }
+
+        [Theory]
+        [InlineData(BuildAction.Clean)]
+        [InlineData(BuildAction.Compile)]
+        [InlineData(BuildAction.Deploy)]
+        [InlineData(BuildAction.Link)]
+        [InlineData(BuildAction.Package)]
+        [InlineData(BuildAction.Rebuild)]
+        public async Task IsUpToDateAsync_False_NonBuildAction(BuildAction buildAction)
+        {
+            await SetupAsync();
+
+            await AssertNotUpToDateAsync(buildAction: buildAction);
+        }
+
+        [Fact]
+        public async Task IsUpToDateAsync_False_BuildTasksActive()
+        {
+            await SetupAsync();
+
+            _isTaskQueueEmpty = false;
+
+            await AssertNotUpToDateAsync(
+                "Critical build tasks are running, not up to date.",
+                "CriticalTasks");
+        }
+
+        [Fact]
+        public async Task IsUpToDateAsync_False_ProjectVersionIncreased()
+        {
+            await SetupAsync();
+
+            _projectVersion++;
+
+            await AssertNotUpToDateAsync(
+                "Project information is older than current project version, not up to date.",
+                "ProjectInfoOutOfDate");
+        }
+
+        [Fact]
+        public async Task IsUpToDateAsync_True_ProjectVersionDecreased()
+        {
+            await SetupAsync();
+
+            _projectVersion--;
+
+            await AssertUpToDate();
+        }
+
+        [Fact]
+        public async Task IsUpToDateAsync_False_ItemsChanged()
+        {
+            await SetupAsync();
+
+            // Add new items
+            BroadcastChange(sourceSnapshot: new Dictionary<string, IProjectRuleSnapshotModel>
+            {
+                ["Content"] = SimpleItems("ItemPath1", "ItemPath2")
+            });
+
+            await AssertNotUpToDateAsync(
+                "The list of source items has changed since the last build, not up to date.",
+                "ItemInfoOutOfDate");
+        }
+
+        [Fact]
+        public async Task IsUpToDateAsync_False_Disabled()
+        {
+            await SetupAsync();
+
+            BroadcastChange(disableFastUpToDateCheck: true);
+
+            await AssertNotUpToDateAsync(
+                "The 'DisableFastUpToDateCheck' property is true, not up to date.",
+                "Disabled");
+        }
+
+        [Fact]
+        public async Task IsUpToDateAsync_False_CopyAlwaysItemExists()
+        {
+            // TODO add a Content or None item as CopyAlways and verify (these are currently excluded by CollectInputs)
```

Comment:
Can you point to the bug?

Reply:
This is what I described in #3937. I am looking to address that in this PR.

Reply:
Yeah what I mean is that we usually point to bugs in "TODO". Not a big thing, don't block on it.

Reply:
I was wrong about that bug anyway.